File: Array/3070. Count Submatrices with Top-Left Element and Sum Less Than k.cpp
/*
3070. Count Submatrices with Top-Left Element and Sum Less Than k
*/

// TC:O(m*n) SC:O(m*n)
class Solution {
    using LL = long long;
public:
    int countSubmatrices(vector<vector<int>>& grid, int k) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<LL>> presum(m+1, vector<LL>(n+1));
        int count = 0;
        for(int i = 1; i <= m; i++){
            for(int j = 1; j <= n; j++){
                presum[i][j] = presum[i-1][j] + presum[i][j-1] - presum[i-1][j-1] + grid[i-1][j-1];
                if(presum[i][j] <= k) count += 1;
            }
        }
        
        return count;
        
       
        
    }
};

/*
將所有包含最左上角的元素 並且sum為 <= k 的數量算出來
*/

File: Array/3070. Count Submatrices with Top-Left Element and Sum Less Than k.py
# TC:O(m*n) SC:O(n)
class Solution:
    def countSubmatrices(self, grid: List[List[int]], k: int) -> int:
        col_sum = [0] * len(grid[0])
        ans = 0
        for row in grid:
            s = 0
            for j, x in enumerate(row):
                col_sum[j] += x
                s += col_sum[j]
                if s > k:
                    break
                ans += 1
        return ans

  # 2D prefix sum
  class Solution:
    def countSubmatrices(self, grid: List[List[int]], k: int) -> int:
        presum = [[0] * (len(grid[0]) + 1) for _ in range(len(grid) + 1)]
        ans = 0
        for i, row in enumerate(grid):
            for j, x in enumerate(row):
                presum[i + 1][j + 1] = presum[i][j + 1] + presum[i + 1][j] - presum[i][j] + x
                if presum[i + 1][j + 1] <= k:
                    ans += 1
        return ans
